misc/multistream-select/src/dialer_select.rs

```diff
@@ -141,10 +150,44 @@ pub struct DialerSelectSeq<R, I: Iterator> {
 }
 
 enum SeqState<R, N> {
-    SendHeader { io: MessageIO<R> },
-    SendProtocol { io: MessageIO<R>, protocol: N },
-    FlushProtocol { io: MessageIO<R>, protocol: N },
-    AwaitProtocol { io: MessageIO<R>, protocol: N },
+    SendHeader {
+        io: MessageIO<R>,
+    },
+
+    // Simultaneous open protocol extension
+    SendSimOpen {
+        io: MessageIO<R>,
+        protocol: Option<N>,
+    },
+    FlushSimOpen {
+        io: MessageIO<R>,
+        protocol: N,
+    },
+    AwaitSimOpen {
+        io: MessageIO<R>,
+        protocol: N,
+    },
+    SimOpenPhase {
+        selection: SimOpenPhase<R>,
+        protocol: N,
+    },
+    Responder {
+        responder: crate::ListenerSelectFuture<R, N>,
+    },
```

Comment:
This feels a bit forced into here and could potentially be a bit more modularised but I think that would come at the cost of a lot of code duplication downstream because we would return different futures?